docs/architecture-decisions/0003-implement-as-a-rest-api.md

```diff
@@ -0,0 +1,60 @@
+
+In early design exercises, we envisioned TACOS more in terms of the staff workflow of analyzing search terms. In the
+longer term, we would also want the tool to be able to evaluate search terms algorithmically.
+
+In either case, the tool should be able to return its evaluation of a search term to a search interface, so that
+interface can better detect user intent. The basic flow might look like this:
+
+1. A user pastes a citation into the Bento input field and executes a search.
+2. Bento asks TACOS (possibly via ActiveRecord callbacks) to analyze the search term.
+3. TACOS evaluates the search term and hopefully identifies it as a citation.
+4. TACOS returns its evaluation to Bento.
+5. In the search results, Bento shares its best guess as to what the user was searching for. In this case, it would be
+a link to the article, but it might also be a database or journal, a library service offering, or contact information
+of a subject matter expert in the libraries.
+
+This pattern would also replace the 'fact' feature built into the TIMDEX UI. In other words, TACOS would be responsible
+for matching DOIs, ISBNS, ISSNs, and other identifiers that could resolve known-item searches.
+
+## GraphQL vs REST
+
+### REST
+
+For a single endpoint API like we are imagining, the simplicity of REST is a nice benefit. We are imagining a minimal parameters and a single structured JSON data response and REST is very clean to use for this purpose.
+
+REST is a ubiquitous architectural style that is more established than GraphQL. If we plan to make TACOS something that
+other organizations can fork and customize for their own needs, then REST will lower the barrier to engage with the
+codebase.
+
+If we choose REST, we should assume we'll need to include a breaking version at some point in the future and start our routes versioned (i.e. `/api/v1?searchterm?=` rather than just `/?searchterm=`).
+
+### GraphQL
+
+GraphQL requires a bit more care to implement from a consumption perspective as it is likely we'd break each portion of the structured JSON data into an optional returned value so consumers would need to explicitly ask for the data they plan to use. This can also be a benefit as we'd only send consumers the data they need, so while there is friction it is fairly small once consumers embrace how GraphQL works.
+
+GraphQL would allow us to have an interactive dashboard, GraphiQL, similar to what we have for TIMDEX to allow for an easy way to see responses from the API in a simple, self documenting web UI.
+
+While we don't anticipate a need for this so early on, GraqphQL's deprecation instead of versioning could come in handy if this API is long lived. We were able to leverage this well in TIMDEX.
+
+Setting up GraphQL enpoints intially would be a bit more work than REST, but we have experience on our team doing this so it is not a blocker.
+
+## Decision
+
+The functionality to return search term evalutions will be implemented as a REST API.
```

Comment:
At the moment I'm not sure that I can come to a position on this ADR. In gradient of agreement terms, I'm not going to block either decision - but I also can't get to a place where I support either choice without further discussion.

Some of the considerations about this choice, for me, are:
- Standardization with our other APIs (where we just retired REST in Timdex API, and am I'm not sure what other APIs we're either maintaining or asking our community to use).
- Whether our architectural principles discuss this question (the latest draft I see doesn't mention either)
- Whether the various uses of this system might need the flexibility offered by GraphQL (I don't have a clear picture of what those uses are right now, so I'm having a hard time getting a meaningful signal here).

Bottom line for me is that more discussion is needed.

Reply:
In gradients of agreement terms, I am probably at "support with reservations" or "agreement with a minor point of contention." I agree with your last bullet point questioning whether GraphQL's flexibility adds value to this use case. I would also suggest that REST's ubiquity could be advantageous to encourage engagement with the codebase, but I'm not sure if that is important in this instance.

Ultimately, neither of these concerns rise to a blocking level, particularly given GraphQL's built-in documentation features and the fact that TIMDEX is now fully GraphQL. API standardization is an interesting question, though, and I'd be curious if we're maintaining other REST APIs that I'm unware of.

Reply:
Interesting. Do we envision non-MITL systems interacting with this API? If no, I don't think REST's ubiquity matters. I can see both sides. This is an internal tool we need to evaluate our understanding of search terms and hopefully return some interesting/useful bits of data about the terms... but what if it was open? Maybe we only send stuff from MITL to our MITL experts for categorization workflows but allow anyone to use the lookup portion of the API once we have some useful algorithms in place?